dll/win32/shell32/folders/CRegFolder.cpp

```diff
@@ -22,110 +22,104 @@
+    LPITEMIDLIST pidl = (LPITEMIDLIST)SHAlloc(cbTotal);
+    if (pidl)
     {
-        /* Tell the caller to run the default action */
-        hr = S_FALSE;
+        ZeroMemory(pidl, cbTotal);
+        pidl->mkid.cb = cb;
+        pidl->mkid.abID[0] = type;
+        pidl->mkid.abID[1] = order;
+        *(CLSID*)(SIZE_T(pidl) + offset) = clsid;
     }
+    return pidl;
+}
 
-    SHFree(pidlFolder);
-    _ILFreeaPidl(apidl, cidl);
-
-    return hr;
+static LPITEMIDLIST CreateRegItem(PIDLTYPE type, LPCWSTR clsidstr)
+{
+    CLSID clsid;
+    return SUCCEEDED(CLSIDFromString(clsidstr, &clsid)) ? CreateRegItem(type, clsid) : NULL;
 }
 
-HRESULT CGuidItemContextMenu_CreateInstance(PCIDLIST_ABSOLUTE pidlFolder,
-                                            HWND hwnd,
-                                            UINT cidl,
-                                            PCUITEMID_CHILD_ARRAY apidl,
-                                            IShellFolder *psf,
-                                            IContextMenu **ppcm)
+HRESULT FormatGUIDKey(LPWSTR KeyName, SIZE_T KeySize, LPCWSTR RegPath, const GUID* riid)
 {
-    HKEY hKeys[10];
-    UINT cKeys = 0;
+    WCHAR xriid[CHARS_IN_GUID];
+    StringFromGUID2(*riid, xriid, _countof(xriid));
+    return StringCchPrintfW(KeyName, KeySize, RegPath, xriid);
+}
 
-    GUID *pGuid = _ILGetGUIDPointer(apidl[0]);
-    if (pGuid)
+static DWORD SHELL_QueryCLSIDValue(_In_ REFCLSID clsid, _In_opt_ LPCWSTR SubKey, _In_opt_ LPCWSTR Value, _In_opt_ PVOID pData, _In_opt_ PDWORD pSize)
+{
+    WCHAR Path[MAX_PATH];
+    wcscpy(Path, L"CLSID\\");
+    StringFromGUID2(clsid, Path + 6, 39);
+    if (SubKey)
     {
-        WCHAR key[60];
-        wcscpy(key, L"CLSID\\");
-        StringFromGUID2(*pGuid, &key[6], 39);
-        AddClassKeyToArray(key, hKeys, &cKeys);
+        wcscpy(Path + 6 + 38, L"\\");
+        wcscpy(Path + 6 + 39, SubKey);
     }
+    return RegGetValueW(HKEY_CLASSES_ROOT, Path, Value, RRF_RT_ANY, NULL, pData, pSize);
+}
 
-    // FIXME: CRegFolder should be aggregated by its outer folder and should
-    // provide the attributes for all required non-registry folders.
-    // It currently does not so we have to ask the outer folder ourself so
-    // that we get the correct attributes for My Computer etc.
-    CComPtr<IShellFolder> pOuterSF;
-    SHBindToObject(NULL, pidlFolder, IID_PPV_ARG(IShellFolder, &pOuterSF));
+static bool HasCLSIDShellFolderValue(REFCLSID clsid, LPCWSTR Value)
+{
+    return SHELL_QueryCLSIDValue(clsid, L"ShellFolder", Value, NULL, NULL) == ERROR_SUCCESS;
+}
 
-    SFGAOF att = (psf && cidl) ? SHGetAttributes(pOuterSF ? pOuterSF.p : psf, apidl[0], SFGAO_FOLDER) : 0;
-    if (att & SFGAO_FOLDER)
-        AddClassKeyToArray(L"Folder", hKeys, &cKeys);
+struct CRegFolderInfo {
+    const REGFOLDERINFO *m_pInfo;
 
-    return CDefFolderMenu_Create2(pidlFolder, hwnd, cidl, apidl, psf, RegFolderContextMenuCallback, cKeys, hKeys, ppcm);
-}
+    void Initialize(const REGFOLDERINFO *pInfo)
```

Comment:
If you give this another name then you don't have to cast in the `Initialize` of all classes implementing this.